### PR TITLE
[v2] Upgrade Solito to Support Next.js 13

### LINKED ISSUE #202: [Done] Next.js 13 support
👀

```sh
yarn add solito
```

Update: available in v2: https://solito.dev/v2

### PR DESCRIPTION
**Pull request by nandorojo**
To try this:

```sh
yarn add solito@next13
```

- Upgrade examples to Next.js 13
- Upgrade examples to use Solito v2 (incomplete)
- Upgrade Reanimated SWC Plugin for Next.js 13

Closes #202

### CODE REVIEW

File: src/link/core.tsx [hunk 1]
```diff
@@ -23,7 +23,7 @@ function LinkCore({
 }) {
   if (Platform.OS === 'web') {
     return (
-      <NextLink {...props} href={href} as={as} passHref>
+      <NextLink {...props} href={href} as={as} passHref legacyBehavior>
```
Comment (xinha-sh) [inline src/link/core.tsx:26]: Instead of legacyBehavior, I would prefer v1. It will be easier to understand which legacy version is being referenced.

Comment (nandorojo) [reply src/link/core.tsx:26]: What do you mean? This prop just ensures that we’re using the RN element as the `a` tag.

Comment (xinha-sh) [reply src/link/core.tsx:26]: @nandorojo I was unable to deduce what the prop **legacyBehavior** does. Hence thought maybe we can have some better prop name

**Comment by chrisarts** [comment]
Did you experiment with the app folder ? I try to reproduce the document initialprops in the RootLayout but no luck, keeping the _document file doest work either.

**Comment by chrisarts** [comment]
Also expo next adapter just upgrate to support next13 take a look:
https://github.com/expo/expo-cli/tree/main/packages/next-adapter#readme

**Comment by ebg1223** [comment]
submitted PR #230 to fix error on expo due to updated nextjs useRouter hook.

**Comment by nandorojo** [comment]
> Did you experiment with the app folder ? I try to reproduce the document initialprops in the RootLayout but no luck, keeping the _document file doest work either.

Not yet. This will require a new approach to injecting styles and such, without `_document`.

File: src/link/core.tsx (hunk 1 above)
Comment (nandorojo) [reply src/link/core.tsx]: This is a Next.js 13 prop that isn’t exposed to Solito users. https://nextjs.org/blog/next-13#breaking-changes

**Comment by macrozone** [comment]
> > Did you experiment with the app folder ? I try to reproduce the document initialprops in the RootLayout but no luck, keeping the _document file doest work either.
> 
> Not yet. This will require a new approach to injecting styles and such, without `_document`.
> 
> If you want to try, see this: https://beta.nextjs.org/docs/styling/css-in-js
> 
> I'm not sure if react-native-web supports this or not.
> 
> I think we'd need to add this to the root layout:
> 
> ```tsx
> import { useServerInsertedHTML } from 'next/navigation';
> import { AppRegistry } from 'react-native'
> 
> // ...
> useServerInsertedHTML(() => {
>      // where does Main come from in Next 13?
>     AppRegistry.registerComponent('Main', () => Main)
>     // @ts-ignore
>     const { getStyleElement } = AppRegistry.getApplication('Main')
>    
>     return <>{getStyleElement()}</>;
>   });
> ```

did you found a solution for that?

**Comment by AwakenedMind** [comment]
Any updates on this

**Comment by nandorojo** [comment]
Sigh, random pings.

I opened a PR at #397 to add app directory support. Maybe one of you can help figure out how to get `useServerInsertedHTML` to work with React Native.

**Comment by Gabrola** [comment]
@nandorojo check the tamagui starter app https://github.com/tamagui/tamagui/blob/master/starters/next-expo-solito/apps/next/app/TamaguiProvider.tsx

**Comment by nandorojo** [comment]
interesting...does `Main` still work here? may as well try 🤷‍♂️

**Comment by nandorojo** [comment]
Screenshot 2023-06-01 at 6 59 56 PM

Wow, can't believe I got this working lol.

**Comment by Gabrola** [comment]
@nandorojo Woot! 🎉

> interesting...does `Main` still work here? may as well try 🤷‍♂️

I think `Main` actually does nothing here and the `registerComponent` function only is only called to trigger the generation of the RN CSS so you get fetch it using `getStyleElement`, pretty sure if you pass it an empty component it will still work. Also, I think this might be unperformant as it would register the app on each re-render.

Can you try this instead and see if it works?

```typescript
  const [reactNativeApp] = useState(() => {
    AppRegistry.registerComponent('Main', () => () => null);
    // @ts-ignore
    return AppRegistry.getApplication('Main');
  });

  useServerInsertedHTML(() => reactNativeApp.getStyleElement());
```

**Comment by chrisarts** [comment]
> @nandorojo Woot! 🎉
> 
> > interesting...does `Main` still work here? may as well try 🤷‍♂️
> 
> I think `Main` actually does nothing here and the `registerComponent` function only is only called to trigger the generation of the RN CSS so you can fetch it using `getStyleElement`, pretty sure if you pass it an empty component it will still work. Also, I think this might be unperformant as it would register the app on each re-render.
> 
> Can you try this instead and see if it works? Follows roughly what's in the next.js docs [here](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#styled-jsx)
> 
> ```ts
>   const [reactNativeApp] = useState(() => {
>     AppRegistry.registerComponent('Main', () => () => null);
>     // @ts-ignore
>     return AppRegistry.getApplication('Main');
>   });
> 
>   useServerInsertedHTML(() => reactNativeApp.getStyleElement());
> ```

Actually yes, Im not even using the app registry and it works normally, just needs to import the .css with RNW reset.

**Pull request merged**